### ISSUE TITLE
[Deliver] - Not able to reject the app version which is in Pending Developer Release

### ISSUE THREAD

**Issue by Thangaraj-M** [fastlane/fastlane#17539]
### New Issue Checklist

- [ ] Updated fastlane to the latest version
- [x ] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x ] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x ] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description

When one version is in Pending Developer release and If set the flag reject_if_possible with true, the existing version is not rejected. There was error occurred before the reject method is called. It failed in the verify version method.

##### Command executed

fastlane upload_to_appstore keychain_pass:****** build_provided:true --verbose

##### Complete output when running fastlane, including the stack trace and command used

<details>
  <pre> 
+----------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------+
|                                                                                            [32mdeliver 2.160.0 Summary[0m                                                                                            |
+----------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------+
| username                                                                   | ********                                                                                                        |
| app_identifier                                                             | ********                                                                                                |
| app_version                                                                | 3.12.0                                                                                                                           |
| build_number                                                               | 12                                                                                                                               |
| platform                                                                   | ios                                                                                                                              |
| edit_live                                                                  | false                                                                                                                            |
| use_live_version                                                           | false                                                                                                                            |
| metadata_path                                                              | /fastlane/metadata/appstore |
| screenshots_path                                                           | ./screenshots                                                                                                                    |
| skip_binary_upload                                                         | true                                                                                                                             |
| skip_screenshots                                                           | true                                                                                                                             |
| skip_metadata                                                              | false                                                                                                                            |
| skip_app_version_update                                                    | false                                                                                                                            |
| force                                                                      | true                                                                                                                             |
| overwrite_screenshots                                                      | false                                                                                                                            |
| submit_for_review                                                          | true                                                                                                                             |
| reject_if_possible                                                         | false                                                                                                                            |
| automatic_release                                                          | false                                                                                                                            |
| auto_release_date                                                          | 1609322400000                                                                                                                    |
| phased_release                                                             | false                                                                                                                            |
| reset_ratings                                                              | false                                                                                                                            |
| submission_information.add_id_info_limits_tracking                         | false                                                                                                                            |
| submission_information.add_id_info_serves_ads                              | false                                                                                                                            |
| submission_information.add_id_info_tracks_action                           | false                                                                                                                            |
| submission_information.add_id_info_tracks_install                          | false                                                                                                                            |
| submission_information.add_id_info_uses_idfa                               | false                                                                                                                            |
| submission_information.content_rights_has_rights                           | false                                                                                                                            |
| submission_information.content_rights_contains_third_party_content         | false                                                                                                                            |
| submission_information.export_compliance_platform                          | false                                                                                                                            |
| submission_information.export_compliance_compliance_required               | false                                                                                                                            |
| submission_information.export_compliance_encryption_updated                | false                                                                                                                            |
| submission_information.export_compliance_app_type                          | false                                                                                                                            |
| submission_information.export_compliance_uses_encryption                   | false                                                                                                                            |
| submission_information.export_compliance_is_exempt                         | false                                                                                                                            |
| submission_information.export_compliance_contains_third_party_cryptography | false                                                                                                                            |
| submission_information.export_compliance_contains_proprietary_cryptography | false                                                                                                                            |
| submission_information.export_compliance_available_on_french_store         | false                                                                                                                            |
| run_precheck_before_submit                                                 | true                                                                                                                             |
| precheck_default_rule_level                                                | warn                                                                                                                             |
| ignore_language_directory_validation                                       | false                                                                                                                            |
| precheck_include_in_app_purchases                                          | true                                                                                                                             |
+----------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------+

INFO [2020-10-31 18:04:16.76]: Making sure the latest version on App Store Connect matches '3.12.0'...
Exception is... The provided entity includes a relationship with an invalid value - You cannot create a new version of the App in the current state.</pre>
</details>

### Environment

<details>
  <pre>[15:39:59]: Get started using a Gemfile for fastlane https://docs.fastlane.tools/getting-started/ios/setup/#use-a-gemfile
[15:40:02]: Generating fastlane environment output, this might take a few seconds...
<details><summary>🚫 fastlane environment 🚫</summary>

### Stack

| Key                         | Value                                         |
| --------------------------- | --------------------------------------------- |
| OS                          | 10.15.6                                       |
| Ruby                        | 2.7.0                                         |
| Bundler?                    | false                                         |
| Git                         | git version 2.24.3 (Apple Git-128)            |
| Installation Source         | ~/.rvm/rubies/ruby-2.7.0/bin/fastlane         |
| Host                        | Mac OS X 10.15.6 (19G2021)                    |
| Ruby Lib Dir                | ~/.rvm/rubies/ruby-2.7.0/lib                  |
| OpenSSL Version             | OpenSSL 1.1.1g  21 Apr 2020                   |
| Is contained                | false                                         |
| Is homebrew                 | false                                         |
| Is installed via Fabric.app | false                                         |
| Xcode Path                  | /Applications/Xcode12.app/Contents/Developer/ |
| Xcode Version               | 12.0                                          |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   | en_US.UTF-8 | ✅ |
| LANGUAGE | en_US.UTF-8 | ✅ |

### fastlane files:

**No Fastfile found**

**No Appfile found**

### fastlane gems

| Gem      | Version | Update-Status      |
| -------- | ------- | ------------------ |
| fastlane | 2.160.0 | 🚫 Update available |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| did_you_mean              | 1.4.0        |
| executable-hooks          | 1.6.0        |
| bundler-unload            | 1.0.2        |
| rubygems-bundler          | 1.4.5        |
| slack-notifier            | 2.3.2        |
| rouge                     | 2.0.7        |
| xcpretty                  | 0.3.0        |
| terminal-notifier         | 2.0.0        |
| unicode-display_width     | 1.7.0        |
| terminal-table            | 1.8.0        |
| plist                     | 3.5.0        |
| CFPropertyList            | 3.0.2        |
| addressable               | 2.7.0        |
| multipart-post            | 2.0.0        |
| word_wrap                 | 1.0.0        |
| tty-screen                | 0.8.1        |
| tty-cursor                | 0.7.1        |
| tty-spinner               | 0.9.3        |
| colored                   | 1.2          |
| highline                  | 1.7.10       |
| commander-fastlane        | 4.4.6        |
| faraday_middleware        | 1.0.0        |
| fastimage                 | 2.2.0        |
| gh_inspector              | 1.1.3        |
| mini_magick               | 4.10.1       |
| rubyzip                   | 2.3.0        |
| security                  | 0.1.3        |
| xcpretty-travis-formatter | 1.0.0        |
| dotenv                    | 2.7.6        |
| bundler                   | 2.1.2        |
| naturally                 | 2.2.0        |
| simctl                    | 1.6.8        |
| uber                      | 0.1.0        |
| declarative               | 0.0.20       |
| declarative-option        | 0.1.0        |
| representable             | 3.0.4        |
| retriable                 | 3.1.2        |
| mini_mime                 | 1.0.2        |
| multi_json                | 1.15.0       |
| signet                    | 0.14.0       |
| httpclient                | 2.8.3        |
| google-api-client         | 0.38.0       |
| uri                       | 0.10.0       |
| nanaimo                   | 0.3.0        |
| colored2                  | 3.1.2        |
| claide                    | 1.0.3        |
| atomos                    | 0.1.3        |
| xcodeproj                 | 1.19.0       |
| public_suffix             | 4.0.6        |
| babosa                    | 1.0.4        |
| excon                     | 0.78.0       |
| unf_ext                   | 0.0.7.7      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20190701 |
| http-cookie               | 1.0.3        |
| ruby2_keywords            | 0.0.2        |
| faraday                   | 1.1.0        |
| faraday-cookie_jar        | 0.0.7        |
| json                      | 2.3.1        |
| jwt                       | 2.2.2        |
| os                        | 1.1.1        |
| memoist                   | 0.16.2       |
| googleauth                | 0.14.0       |
| rake                      | 13.0.1       |
| digest-crc                | 0.6.1        |
| google-cloud-errors       | 1.0.1        |
| google-cloud-env          | 1.4.0        |
| google-cloud-core         | 1.5.0        |
| google-cloud-storage      | 1.29.1       |
| emoji_regex               | 3.2.0        |
| aws-eventstream           | 1.1.0        |
| aws-sigv4                 | 1.2.2        |
| aws-partitions            | 1.388.0      |
| jmespath                  | 1.4.0        |
| aws-sdk-core              | 3.109.1      |
| aws-sdk-kms               | 1.39.0       |
| aws-sdk-s3                | 1.83.1       |
| forwardable               | 1.3.1        |
| logger                    | 1.4.2        |
| cgi                       | 0.1.0        |
| timeout                   | 0.1.0        |
| stringio                  | 0.1.0        |
| ipaddr                    | 1.2.2        |
| openssl                   | 2.1.2        |
| ostruct                   | 0.2.0        |
| strscan                   | 1.0.3        |
| date                      | 3.0.0        |
| delegate                  | 0.1.0        |
| fileutils                 | 1.4.1        |
| io-console                | 0.5.3        |
| zlib                      | 1.1.0        |
| singleton                 | 0.1.0        |
| rexml                     | 3.2.3        |
| open3                     | 0.1.0        |
| yaml                      | 0.1.0        |
| psych                     | 3.1.0        |
| mutex_m                   | 0.1.0        |
| webrick                   | 1.6.0        |
</details>
</pre>
</details>

**Comment by Thangaraj-M** [comment]
@joshdholtz  Can you please check this issue?

**Comment by Thangaraj-M** [comment]
**Spaceship logs:**
INFO  [18:04:16]: >> GET apps/1529770189/appStoreVersions: [undefined body] 
DEBUG [18:04:18]: << GET apps/1529770189/appStoreVersions: 200 {"data"=>[], "links"=>{"self"=>"https://appstoreconnect.apple.com/iris/v1/apps/1529770189/appStoreVersions?include=build%2CappStoreVersionSubmission&filter%5Bplatform%5D=IOS&filter%5BappStoreState%5D=DEVELOPER_REJECTED%2CINVALID_BINARY%2CMETADATA_REJECTED%2CPREPARE_FOR_SUBMISSION%2CREJECTED%2CWAITING_FOR_REVIEW"}, "meta"=>{"paging"=>{"total"=>0, "limit"=>50}}}
INFO  [18:04:18]: >> POST appStoreVersions: {"data":{"type":"appStoreVersions","attributes":{"versionString":"3.12.0","platform":"IOS"},"relationships":{"app":{"data":{"type":"apps","id":"1529770189"}}}}} 
DEBUG [18:04:19]: << POST appStoreVersions: 409 {"errors"=>[{"id"=>"0b344ffd-bd9c-41de-81f7-586f14cdb489", "status"=>"409", "code"=>"ENTITY_ERROR.RELATIONSHIP.INVALID", "title"=>"The provided entity includes a relationship with an invalid value", "detail"=>"You cannot create a new version of the App in the current state.", "source"=>{"pointer"=>"/data/relationships/app"}}]}

**Comment by max-ott** [comment]
@Thangaraj-M Please don't `@` contributors, this is just rude. Could you please update to the latest fastlane version first?

**Comment by Thangaraj-M** [comment]
I have upgraded the fastlane version 2.165.0 and tried. I have set the flag **reject_if_possible** with true.But still I am getting the same error. 

+---------------------------------------------------------------------+---------------------------------------------------------------------+
|                                                          deliver 2.165.0 Summary                                                          |
+---------------------------------------------------------------------+---------------------------------------------------------------------+
| username                                                            | *******                                          |
| app_identifier                                                      | *******                                  |
| app_version                                                         | 3.12.0                                                              |
| build_number                                                        | 12                                                                  |
| platform                                                            | ios                                                                 |
| edit_live                                                           | false                                                               |
| use_live_version                                                    | false                                                               |
| metadata_path                                                       | fastlane/metadata/  |
|                                                                     | appstore                                                            |
| screenshots_path                                                    | ./screenshots                                                       |
| skip_binary_upload                                                  | true                                                                |
| skip_screenshots                                                    | true                                                                |
| skip_metadata                                                       | false                                                               |
| skip_app_version_update                                             | false                                                               |
| force                                                               | true                                                                |
| overwrite_screenshots                                               | false                                                               |
| submit_for_review                                                   | true                                                                |
**| reject_if_possible                                                  | true                                                                |**
| auto_release_date                                                   | 1606384800000                                                       |
| phased_release                                                      | false                                                               |
| reset_ratings                                                       | false                                                               |
| submission_information.add_id_info_limits_tracking                  | false                                                               |
| submission_information.add_id_info_serves_ads                       | false                                                               |
| submission_information.add_id_info_tracks_action                    | false                                                               |
| submission_information.add_id_info_tracks_install                   | false                                                               |
| submission_information.add_id_info_uses_idfa                        | false                                                               |
| submission_information.content_rights_has_rights                    | false                                                               |
| submission_information.content_rights_contains_third_party_content  | false                                                               |
| submission_information.export_compliance_platform                   | false                                                               |
| submission_information.export_compliance_compliance_required        | false                                                               |
| submission_information.export_compliance_encryption_updated         | false                                                               |
| submission_information.export_compliance_app_type                   | false                                                               |
| submission_information.export_compliance_uses_encryption            | false                                                               |
| submission_information.export_compliance_is_exempt                  | false                                                               |
| submission_information.export_compliance_contains_third_party_cryp  | false                                                               |
| tography                                                            |                                                                     |
| submission_information.export_compliance_contains_proprietary_cryp  | false                                                               |
| tography                                                            |                                                                     |
| submission_information.export_compliance_available_on_french_store  | false                                                               |
| run_precheck_before_submit                                          | true                                                                |
| precheck_default_rule_level                                         | warn                                                                |
| ignore_language_directory_validation                                | false                                                               |
| precheck_include_in_app_purchases                                   | true                                                                |
+---------------------------------------------------------------------+---------------------------------------------------------------------+

INFO [2020-11-02 19:37:01.88]: Making sure the latest version on App Store Connect matches '3.12.0'...
Exception is... The provided entity includes a relationship with an invalid value - You cannot create a new version of the App in the current state.

**Comment by Thangaraj-M** [comment]
In the deliver runner.rb file, inside the run method, verify_version method is invoked before the reject_version_if_possible method.
At the time of verifying the version, if no editable version present in the Appstore connect, it trying to create new version and it failed.
![reject](https://user-images.githubusercontent.com/65262635/97884415-0eec0180-1d4c-11eb-98e3-06889858f907.png)

**Comment by max-ott** [comment]
I think there's a logic error in your `verify` part. I seems the script is trying to create a new version, when this is not possible.

**Comment by Thangaraj-M** [comment]
This the Fastlane library code. I have taken from that only.

**Comment by ainame** [comment]
Hi. I tried to interpret your error. Although I couldn't find backtrace logs you should have shared here, I assumed you got the error at this line. 
https://github.com/fastlane/fastlane/blob/7add7e9432c351a144809023013bb7ff6c529daf/deliver/lib/deliver/runner.rb#L120

> DEBUG [18:04:19]: << POST appStoreVersions: 409 {"errors"=>[{"id"=>"0b344ffd-bd9c-41de-81f7-586f14cdb489", "status"=>"409", "code"=>"ENTITY_ERROR.RELATIONSHIP.INVALID", "title"=>"The provided entity includes a relationship with an invalid value", "detail"=>"You cannot create a new version of the App in the current state.", "source"=>{"pointer"=>"/data/relationships/app"}}]}

This log, specifically 409 status code, presumably means that you can't create another "App Store Version" with the same version number with `POST /v1/appStoreVersions` endpoint. 
https://developer.apple.com/documentation/appstoreconnectapi/create_an_app_store_version

To dig deep, `app.ensure_version!` internally tries creating an app version if it doesn't find the existing app version by the query `get_edit_app_store_version`. However `get_edit_app_store_version` doesn't include "Pending Developer Release" status after all so that the conflict error can occur in your case.

https://github.com/fastlane/fastlane/blob/7add7e9432c351a144809023013bb7ff6c529daf/spaceship/lib/spaceship/connect_api/models/app.rb#L187-L202

https://github.com/fastlane/fastlane/blob/7add7e9432c351a144809023013bb7ff6c529daf/spaceship/lib/spaceship/connect_api/models/app.rb#L226-L244

So I feel like `app.ensure_version!` should also check its version with ones returned by `get_pending_release_app_store_version`🤔  

Possibly, I can submit a PR for this but can't test it by myself since I have no app being "Pending Developer Release" status...

**Comment by Thangaraj-M** [comment]
Thank you for your analysis. I want to reject the build which is in Pending Developer Release and submit a new version. The version creation happened before the rejection. So the new version creation is failed. I think the order need to be change with the rejection should be done before getting the editable versions.

**Comment by ainame** [comment]
You don't need to swap the order. `app.ensure_version!` works as `find or create` with a given version string. It's just missing the app version being "Pending Developer Release" state right now.

**Comment by Thangaraj-M** [comment]
An app with one version 1.2.1 is present in Pending Developer Release state. Now trying to submit new build with another version 3.10.12. If the app is in Pending Developer Release state, we can not edit the version. To edit the version, need to reject that first. Then the state will be changed to Developer Rejected and the version can be editable. 
If we add the Pending Developer Release state also in the ensure version, then the version value will not be updated with new one.

**Comment by ainame** [comment]
> An app with one version 1.2.1 is present in Pending Developer Release state. Now trying to submit new build with another version 3.10.12. If the app is in Pending Developer Release state, we can not edit the version. To edit the version, need to reject that first. Then the state will be changed to Developer Rejected and the version can be editable.
If we add the Pending Developer Release state also in the ensure version, then the version value will not be updated with new one.

Ah, I understood the point👍 That makes sense. I don't know how come it used to work previously though🤔

**Comment by ainame** [comment]
Apparently, it hasn't been working for a while since at some point.

https://github.com/fastlane/fastlane/issues/14154

**Comment by joshdholtz** [comment]
Will look into!

**Comment by tastafur** [comment]
So have you tested if this is the solution? "reject_if_possible before verify_version"

**Comment by ainame** [comment]
```
Spaceship::UnexpectedResponse: [!] The provided entity includes a relationship with an invalid value - You cannot create a new version of the App in the current state.
The provided entity includes an attribute with a value that has already been used - The version number has been previously used.
  /path/to/fastlane-2.168.0/spaceship/lib/spaceship/connect_api/api_client.rb:193:in `handle_response'
  /path/to/fastlane-2.168.0/spaceship/lib/spaceship/connect_api/api_client.rb:124:in `post'
  /path/to/fastlane-2.168.0/spaceship/lib/spaceship/connect_api/tunes/tunes.rb:768:in `post_app_store_version'
  /path/to/.rbenv/versions/2.6.6/lib/ruby/2.6.0/forwardable.rb:230:in `post_app_store_version'
  /path/to/fastlane-2.168.0/spaceship/lib/spaceship/connect_api/models/app.rb:210:in `ensure_version!'
  /path/to/fastlane-2.168.0/deliver/lib/deliver/runner.rb:120:in `verify_version'
  /path/to/fastlane-2.168.0/deliver/lib/deliver/runner.rb:48:in `run'
```

I've got the above error within `ensure_version!` when an app had "Pending Developer Release" state and we tried to resubmit it without bumping a version whilst the original report was trying to bump a version. This is exactly what I expected outcome with the current code at the above comment😞 

https://github.com/fastlane/fastlane/issues/17539#issuecomment-720875693

**Comment by kambala-decapitator** [comment]
FYI this is how I reject apps that are pending release:

```ruby
    Spaceship::ConnectAPI.login(
      use_portal: false,
      tunes_team_id: teamID,
    )

    app = Spaceship::ConnectAPI::App.find(bundleID)
    if app.nil?
      UI.important "not found in team #{teamID} on ASC!"
      next
    end

    if !app.reject_version_if_possible!(platform: Spaceship::ConnectAPI::Platform::IOS)
      UI.important "couldn't reject!"
      next
    end
    UI.success "rejected"
```

**Comment by tastafur** [comment]
> So have you tested if this is the solution? "reject_if_possible before verify_version"

I had a pending release and it failed me even with reject_if_possible, I now connect through App Store Connect Api

**Comment by andrew54068** [comment]
Here is a plugin I wrote to cancel or release, might help for those encounter the same situation.
you can call this plugin action before Deliver
https://github.com/andrew54068/fastlane-plugin-taiwan_number_one

**Comment by stherold** [comment]
any news on this ? would be nice if it "just works" without additional plugin

**Comment by kambala-decapitator** [comment]
@stherold see my solution above.

If you want to do it with a single command, I guess you'd have to make a PR for that.

**Comment by kg-currenxie** [comment]
> > So have you tested if this is the solution? "reject_if_possible before verify_version"
> 
> I had a pending release and it failed me even with reject_if_possible, I now connect through App Store Connect Api

Same issue here. `reject_if_possible` true.
I even tried to upload a new version. Current pending version 1.2.6, but uploading 1.2.7. Same error.

**Comment by joshdholtz** [comment]
Will look into! I think I know the cause of this one 🤷‍♂️

**Comment by sneko** [comment]
I also got the error:
```
The provided entity includes a relationship with an invalid value - You cannot create a new version of the App in the current state.
```

and the only way was to manually cancel the pending reviewed release from the Apple UI. A bit annoying but that's fine 👍  hope a fix will be done so the `reject_if_possible` can work.

Thanks!

**Comment by icedice** [comment]
Just ran into this issue too. I have 20 apps in "pending" state that I need to cancel, so I too am hoping for a fix.

**Comment by kambala-decapitator** [comment]
@icedice until this feature isn't implemented as a single command / param in fastlane, you can use the solution given in the above comment: https://github.com/fastlane/fastlane/issues/17539#issuecomment-749059538

**Comment by icedice** [comment]
Oh, thank you. Not sure how I missed that.

**Comment by alimcharaniya** [comment]
this is still an issue for me

**Comment by MantasSkara** [comment]
> FYI this is how I reject apps that are pending release:
> 
> ```ruby
>     Spaceship::ConnectAPI.login(
>       use_portal: false,
>       tunes_team_id: teamID,
>     )
> 
>     app = Spaceship::ConnectAPI::App.find(bundleID)
>     if app.nil?
>       UI.important "not found in team #{teamID} on ASC!"
>       next
>     end
> 
>     if !app.reject_version_if_possible!(platform: Spaceship::ConnectAPI::Platform::IOS)
>       UI.important "couldn't reject!"
>       next
>     end
>     UI.success "rejected"
> ```

I applied this workaround to my Fastfile, but it doesn't work. I simply added this to the new lane before delivering. Is there anything additionally that should be done?

**Comment by kambala-decapitator** [comment]
@MantasSkara 

> but it doesn't work

which error appeared?

Obviously you need to assign `teamID` and `bundleID` variables to proper values.

**Comment by MantasSkara** [comment]
@kambala-decapitator 
Yeah, sorry for not including details. I hardcoded "bundleID" value here: `Spaceship::ConnectAPI::App.find("myappbundleid")`

As far as the teamID goes, I connect with _Spaceship::ConnectAPI::Token.create_.

The error that I get is _"couldn't reject"_

**Comment by kambala-decapitator** [comment]
@MantasSkara well, you'll receive this error if app is not in a "pending release" state. From your words

> I simply added this to the new lane before delivering

it seems that this is always called regardless of the state.

**Comment by MantasSkara** [comment]
@kambala-decapitator Thanks! I'l try to investigate what is going on.  The app should be in the state. If I go into the app store CONNECT, I see: ### **VERSION** Pending Developer Release for the app.

**Comment by kambala-decapitator** [comment]
@MantasSkara strange, this approach used to work well... Maybe your app is macOS/tvOS? :)

You can also test this code in `irb` where you can interactively inspect variables (like check full state of `app`), maybe you'll be able to find some hint there.

**Comment by Blackjacx** [comment]
We could wait a little bit more and just use the new possibility of cancellation for submissions: 
https://github.com/fastlane/fastlane/pull/19751
